### PR TITLE
Improve Brush node

### PR DESCRIPTION
**Pull request by malpas**
Hi, Hesiod looks great. I thought the brush node felt a bit placeholder, so I've reworked it and fixed a couple UI bugs in the process.

## Changes
- Change the brush node to be more like painting
- Fix ImGui state related bugs

## Screenshot
Capture

### CODE REVIEW

**Comment by otto-link** [comment]
Hey, thanks a lot for the feedback and for your contributions!
Looks really nice, I'll handle your PR pretty quickly

**Review by otto-link — APPROVED** [review]
Hey, thanks again for the PR, I proposed some modifications in the review. Let me know if you have questions.

**Comment by malpas** [comment]
Thanks for the feedback. Looks like the error was for std::span, which compiled on MSVC without needing \<span\>.

Those definitions aren't for bulletproofing the constructor. I see now there's a deeper issue that ViewNode doesn't have a virtual destructor, so the drawing texture may leak when ViewBrush destructs polymorphically. 

Defining the virtual destructor for ViewNode will remove its default move construction/assignment, at which point you'll need to define move and copy semantics for it  (rule of 5), same with ViewBrush. The compiler warnings were from me wrongly defaulting ViewBrush's move constructor.

I agree some of this gets outside the scope of the brush node, unless you feel strongly about the occasional texture leak for now. It might be worthwhile for you to take some time to decide explicitly which of your classes are unique vs shared, and how they're destroyed. Many of those decisions are the kind of top-down changes that I believe would be best for you to make yourself. If nodes become clonable in the future, you might also decide later to add a virtual clone() function.

File: Hesiod/src/nodes/view_node/view_brush.cpp [hunk 1]
```diff
@@ -43,9 +85,52 @@ bool ViewBrush::render_settings()
...
+
   has_changed |= this->render_settings_footer();
 
   return has_changed;
 }
 
+void ViewBrush::sync_drawing_texture()
+{
+  auto array = this->edit_state.pending_hm.to_array();
+  std::vector<uint8_t> img(array.shape.x * array.shape.y);
```
[the remark this answers is not in the document]
Comment (malpas) [reply Hesiod/src/nodes/view_node/view_brush.cpp:105]: This was because colorize grayscale remapped the colours on the canvas

**Comment by otto-link** [comment]
I'm going to merge your PR as soon as I can. The resulting user experience feels really nice, especially comparing to the first mockup 😅  I'm about to integrate a new version of the "noise Node" (like fractal) with the ability to feed the fractal layering with an input heightmap (so that we'll be able to "draw" the fractal noise).

Let me know if there are some specific topics you'd like to work on!

For the Brush node, there are some new features that could be added:
 - substract kernel (with right mouse click?)
 - choose kernel (there are some kernels available in HighMap, there could be a listbox widget to choose among them)
 - node serialization: requires some thinking, saving the heightmap data as ASCII in the json might not be wise... probably have to point to an image file

For the virtual destructor topic, first thanks for spotting the issue. Having an external perspective is really valuable. I'm going to create a specific issue for that with some details on the overall class structures. I'll be glad to have your input on the pros/cons and what could be done to improve it.

**Comment by malpas** [comment]
I appreciate the kind words and will keep an eye out for anything that catches my interest. Right click subtraction should already work.

When it comes to user-facing terrain software, most people just want features to fit together nicely and feel satisfying to create realistic landscapes. We want a mountain here, river there, canyon, etc. It's a very tricky balance between power and UX. Looking forward to seeing what you create in the future.

**Pull request merged**